src/abstractdataframe/abstractdataframe.jl

```diff
@@ -2525,39 +2525,95 @@ end
 
 function _permutation_helper!(fun::Union{typeof(Base.permute!!), typeof(Base.invpermute!!)},
                               df::AbstractDataFrame, p::AbstractVector{<:Integer})
-    toskip = Set{Int}()
+    nrow(df) != length(p) && throw(DimensionMismatch("Permutation does not have a correct length ($(nrow(df)) != $(length(p)))"))
+    
+    cp = _compile_permutation!(Base.copymutable(p))
+
+    isempty(cp) && return df
+
+    if fun === Base.invpermute!! 
+        reverse!(@view cp[1:end-1])
+    end
+
     seen_cols = IdDict{Any, Nothing}()
     for (i, col) in enumerate(eachcol(df))
-        if haskey(seen_cols, col)
-            push!(toskip, i)
-        else
+        if !haskey(seen_cols, col)
             seen_cols[col] = nothing
-        end
-        # p might be a column of df so we make sure we unalias
-        if col === p
-            p = copy(p)
+            _cycle_permute!(col, cp)
         end
     end
+    
+    return df
+end
 
-    pp = similar(p)
+# convert a classical permutation to zero terminated cycle 
+# notation, zeroing the original permutation in the process.
+function _compile_permutation!(p::AbstractVector{<:Integer})
+    firstindex(p) == 1 || 
+        throw(ArgumentError("Permutation vectors must have 1-based indexing"))
+    # this length is sufficient because we do not record 1-cycles,
+    # so the worst case is all 2-cycles. One extra element gives the
+    # algorithm leeway to defer error detection without unsafe reads.
+    # trace _compile_permutation!([3,3,1]) for example.
+    out = similar(p, 3 * length(p) ÷ 2 + 1)
+    out_len = 0
+    start = 0
+    count = length(p)
+    @inbounds while count > 0
+        start = findnext(!iszero, p, start + 1)
+        start isa Int || throw(ArgumentError("Not a permutation"))
```

Comment:
This may not be super explicit for somebody calling `permute!`. Maybe say something like "Passed vector `p` is not a valid permutation"?

Reply:
I have incorporated all your suggestions in the PR. Once the CI passes I will merge it.